Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -1070,20 +1070,26 @@ void TextureMapperLayer::paintWith3DRenderingContext(TextureMapperPaintOptions&
     });
 }
 
-void TextureMapperLayer::collect3DSceneLayers(Vector<TextureMapperLayer*>& layers)
+void TextureMapperLayer::collect3DRenderingContextLayers(Vector<TextureMapperLayer*>& layers)
 {
-    bool isLeafOf3DScene = !m_state.preserves3D && (m_parent && m_parent->preserves3D());
-    if (preserves3D() || isLeafOf3DScene) {
-        if (m_state.visible)
+    bool isLeafOf3DRenderingContext = !m_state.preserves3D && (m_parent && m_parent->preserves3D());
+    if (preserves3D() || isLeafOf3DRenderingContext) {
+        bool hasVisualContent = m_backingStore || m_contentsLayer
+            || (m_state.backgroundColor.isValid() && m_state.backgroundColor.isVisible())
+            || (m_state.solidColor.isValid() && m_state.solidColor.isVisible())
+            || hasFilters() || hasMask() || hasBackdrop();
+
+        // Add layers to 3d rendering context only if they get actually painted.
+        if (m_state.visible && (hasVisualContent || (isLeafOf3DRenderingContext && m_children.size() > 0)))
```

Comment:
I think `(isLeafOf3DRenderingContext && m_children.size() > 0)` can be replaced with `isFlattened()`. Right?

Reply:
No, the specification states: "The element establishing the 3D rendering context, and each other 3D transformed element participating in the 3D rendering context, is rendered into its own plane." This means that if a leaf element has no 3D transformation, it should not be flattened. TextureMapper follows this logic.

Additionally, the condition `(isLeafOf3DRenderingContext && m_children.size() > 0)` addresses cases where `RenderLayerBacking` inserts a zero-sized, non-3D transformed layer into a leaf node. This occurs in `RenderLayerBacking::updateTransformFlatteningLayer`. This layer usually has visible children. Also this condition is necessary for situations like iframe leaf nodes. These cases are on my to-do list for further investigation. However, I think it is more critical to address the performance regression with the posterl-circles first.

Reply:
Got it. `m_children.size() > 0)` should be `!m_children.isEmpty()`.

Reply:
Fixed